android-components/components/feature/fxsuggest/src/main/java/mozilla/components/feature/fxsuggest/FxSuggestStorage.kt

```diff
@@ -35,25 +43,37 @@ class FxSuggestStorage(
     private val readScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
     private val writeScope: CoroutineScope = CoroutineScope(Dispatchers.IO)
 
+    private val logger = Logger("FxSuggestStorage")
+
     /**
      * Queries the store for suggestions.
      *
      * @param query The input and suggestion types to match.
      * @return A list of matching suggestions.
      */
     suspend fun query(query: SuggestionQuery): List<Suggestion> =
-        withContext(readScope.coroutineContext) {
-            store.value.query(query)
+        handleSuggestExceptions("query", emptyList()) {
+            withContext(readScope.coroutineContext) {
+                store.value.query(query)
+            }
         }
 
     /**
      * Downloads and persists new Firefox Suggest search suggestions.
      *
      * @param constraints Optional limits on suggestions to ingest.
+     * @return `true` if ingestion succeeded; `false` if ingestion failed and should be retried.
      */
-    suspend fun ingest(constraints: SuggestIngestionConstraints = SuggestIngestionConstraints()) =
-        withContext(writeScope.coroutineContext) {
-            store.value.ingest(constraints)
+    suspend fun ingest(constraints: SuggestIngestionConstraints = SuggestIngestionConstraints()): Boolean =
+        handleSuggestExceptions("ingest", false) {
+            withContext(writeScope.coroutineContext) {
+                try {
+                    store.value.ingest(constraints)
+                    true
+                } catch (e: SuggestApiException) {
+                    false
+                }
+            }
         }
```

Comment:
```suggestion
        withContext(writeScope.coroutineContext) {
            handleSuggestExceptions("ingest", false) {
                store.value.ingest(constraints)
                true
            }
        }
```

If I'm not mistaken, you can flip these nested methods and then you don't need a second try-catch because the `handleSuggestExceptions` already has this in there.

Reply:
In fact, I think we would end up dropping the exception and not hitting the crash reporter because we return false on line 74. 🤔

Reply:
> I think we would end up dropping the exception and not hitting the crash reporter because we return false on line 74.

I did this intentionally—my thinking was, `ingest()` is a background operation that can fail for any number of reasons, like connectivity issues, and we don't necessarily want to report all of them—but I actually like your suggestion better!

We have a ticket on file for the Rust component to improve the exception hierarchy—right now, everything's a `SuggestApiException.Other`, so it doesn't distinguish between HTTP errors, SQLite errors, interruptions—but until then, maybe it's better to report all of them for now.

WDYT?

Reply:
Yeah, it's fine reporting them all for now. We've done this in the past with places and push rust components. Those errors are reported as "infos" so they don't trigger any crash alerts. Eventually, we were able to ignore the noisy ones that were not actionable and only report the valuable ones.

_Narrator: why even throw if we are going to ignore them? 👀_